File: Frontend/mobile_app/firefighter_safety_data_hub/lib/models/ppe_model.dart
class PPE {
  final String? eventId;
  final String? helmetId;
  final String? hoodId;
  final String? faceMaskId;
  final String? scbaId;
  final String? gloveId;
  final String? bootId;
  final String? turnoutCoat;
  final String? turnoutPants;
  final DateTime lastUpdated;

  PPE({
    this.eventId,
    this.helmetId,
    this.hoodId,
    this.faceMaskId,
    this.scbaId,
    this.gloveId,
    this.bootId,
    this.turnoutCoat,
    this.turnoutPants,
    required this.lastUpdated,
  });

  Map<String, dynamic> toApiJson({bool isPpeUpdated = false}) {
    String s(String? v) => (v == null || v.trim().isEmpty) ? '' : v.trim();

    return {
      if (eventId != null && eventId!.trim().isNotEmpty)
        'event_id': eventId!.trim(),
      'helmet_id': s(helmetId),
      'hood_id': s(hoodId),
      'face_mask_id': s(faceMaskId),
      'scba_id': s(scbaId),
      'glove_id': s(gloveId),
      'boot_id': s(bootId),
      'bunker_coat_id': s(turnoutCoat),
      'bunker_pants_id': s(turnoutPants),
      'is_ppe_updated': isPpeUpdated,
    };
  }

  Map<String, dynamic> toJson() {
    return {
      'eventId': eventId,
      'helmetId': helmetId,
      'hoodId': hoodId,
      'faceMaskId': faceMaskId,
      'scbaId': scbaId,
      'gloveId': gloveId,
      'bootId': bootId,
      'turnoutCoat': turnoutCoat,
      'turnoutPants': turnoutPants,
      'lastUpdated': lastUpdated.toIso8601String(),
    };
  }

  factory PPE.fromJson(Map<String, dynamic> json) {
    return PPE(
      eventId: json['eventId'],
      helmetId: json['helmetId'],
      hoodId: json['hoodId'],
      faceMaskId: json['faceMaskId'],
      scbaId: json['scbaId'],
      gloveId: json['gloveId'],
      bootId: json['bootId'],
      turnoutCoat: json['turnoutCoat'],
      turnoutPants: json['turnoutPants'],
      lastUpdated: DateTime.parse(json['lastUpdated']),
    );
  }
}

File: Backend/app/models/PpeModel.py
import uuid
from typing import Optional

from sqlalchemy import Boolean, DateTime, ForeignKey, String, func
from sqlalchemy.dialects.postgresql import UUID
from sqlalchemy.orm import Mapped, mapped_column
from sqlalchemy import String

from app.db.base import Base


class PpeModel(Base):
    __tablename__ = "ppe_table"

    ppe_id: Mapped[uuid.UUID] = mapped_column(UUID(as_uuid=True), primary_key=True, default=uuid.uuid4)
    user_id: Mapped[str] = mapped_column(String, ForeignKey("users.id", ondelete="CASCADE"), nullable=False)
    event_id: Mapped[Optional[uuid.UUID]] = mapped_column(UUID(as_uuid=True), ForeignKey("log_events.event_id", ondelete="SET NULL"), nullable=True)

    helmet_id: Mapped[str] = mapped_column(String(255), nullable=False, default="")
    hood_id: Mapped[str] = mapped_column(String(255), nullable=False, default="")
    face_mask_id: Mapped[str] = mapped_column(String(255), nullable=False, default="")
    scba_id: Mapped[str] = mapped_column(String(255), nullable=False, default="")
    glove_id: Mapped[str] = mapped_column(String(255), nullable=False, default="")
    boot_id: Mapped[str] = mapped_column(String(255), nullable=False, default="")
    bunker_coat_id: Mapped[str] = mapped_column(String(255), nullable=False, default="")
    bunker_pants_id: Mapped[str] = mapped_column(String(255), nullable=False, default="")

    is_ppe_updated: Mapped[bool] = mapped_column(Boolean, nullable=False, default=False)
    created_at: Mapped[str] = mapped_column(DateTime(timezone=True), server_default=func.now(), nullable=False)